sbom/examples/container_image/from_catalog.py

```diff
@@ -0,0 +1,229 @@
+import json
+import sys
+
+import requests
+
+# These container images (identified by their NVR) are known to contain only RPM packages and no
+# other content type.
+RPM_CONTAINER_IMAGES = [
+    "ubi9-micro-container-9.4-6.1716471860",
+    "podman-container-9.4-8",
+    "kernel-module-management-operator-container-1.1.2-25",
+]
+
+catalog_url = "https://catalog.redhat.com/api/containers/v1/"
+nvr_api = catalog_url + "images/nvr/"
+rpm_manifest_api = catalog_url + "images/id/{catalog_image_id}/rpm-manifest"
+
+rpm_sbom_url = "https://access.redhat.com/security/data/sbom/v1/rpm/"
+
+
+def get_image_data(image_nvr):
+    response = requests.get(nvr_api + image_nvr)
+    response.raise_for_status()
+    # This is a paged response, but we're assuming there are not 100+ images for a single
+    # container image NVR.
+    yield from response.json()["data"]
+
+
+def get_rpms(image_id):
+    response = requests.get(rpm_manifest_api.format(catalog_image_id=image_id))
+    response.raise_for_status()
+    yield from response.json()["rpms"]
+
+
+def generate_sbom_for_image(image_nvr):
+    # Split to e.g. "ubi9-micro-container" and "9.4-6.1716471860"
+    image_nvr_name, *image_nvr_version = image_nvr.rsplit("-", maxsplit=2)
+    image_nvr_version = "-".join(image_nvr_version)
+
+    packages = []
+    relationships = []
+    image_index_pkg_created = False
+
+    for image in get_image_data(image_nvr):
+        catalog_image_id = image["_id"]
+        image_digest = image["image_id"]
+        content_sets = image["content_sets"]
+
+        # A container image may be available through more than one repo; collect all repos,
+        # registries they are available from, and the most specific tag for each repo image.
+        repos = set()
+        image_index_digest = ""
+        for repo in image["repositories"]:
+            registry = repo["registry"]
+            repo_name = repo["repository"]
+            repo_namespace, _, repo_name = repo_name.rpartition("/")
+            if repo_namespace:
+                registry = f"{registry}/{repo_namespace}"
```

Comment:
I think the repository_url is not intended to be the namespace but the full registry/namespace/name part. See the examples here: https://github.com/package-url/purl-spec/blob/master/PURL-TYPES.rst

So for example, I think we want output like this:
```
"externalRefs": [
  {
    "referenceCategory": "PACKAGE-MANAGER",
    "referenceType": "purl",
    "referenceLocator": "pkg:oci/podman@sha256%3A749dc1a252dfec536ae15bbb4f509a3bccb40de6abf9712619cb8df2d99e94b7?repository_url=registry.access.redhat.com/rhel9/podman&tag=9.4-8"
  },
```

Reply:
Mmm, you're right, nice catch! I also need to fix that in the purl guidelines doc.

Reply:
https://github.com/RedHatProductSecurity/security-data-guidelines/pull/7